File: nodejs/test/approved_fixes_poll_contract.test.js
/**
 * Contract test — dashboard-approved fix polling in node_agent.js.
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const path = require('path');

test('node_agent exposes processApprovedFixes and polls approved+applying', () => {
  const source = fs.readFileSync(path.join(__dirname, '..', 'node_agent.js'), 'utf8');
  assert.ok(source.includes('processApprovedFixes'), 'processApprovedFixes missing');
  assert.ok(source.includes('applyApprovedError'), 'applyApprovedError missing');
  assert.ok(source.includes('APPROVED_APPLY_IN_FLIGHT'), 'in-flight guard missing');
  assert.ok(source.includes("'approved', 'applying'"), 'must poll approved and applying');
  assert.ok(source.includes('processApprovedFixes().catch'), 'monitor loop must schedule poll');
  // advanced_agent_testing: approved apply must report test/results
  const applyParts = source.split('async function applyApprovedError');
  assert.ok(applyParts.length > 1, 'applyApprovedError missing');
  const applyFn = applyParts[1].split('async function discoverApiUrl')[0];
  assert.ok(
    applyFn.includes('await runTestsAndReport(errorId, applyResult.success)'),
    'approved apply must call runTestsAndReport',
  );
  // post-apply-config/connector is unsigned — must not verify response HMAC
  const paParts = source.split('async function getPostApplyConnectorJson');
  assert.ok(paParts.length > 1, 'getPostApplyConnectorJson missing');
  const paFn = paParts[1].split('async function runPostApplySteps')[0];
  assert.ok(!paFn.includes('verifyResponseHmac'), 'post-apply config must not require response HMAC');
  const exportsBlock = require('../node_agent.js');
  assert.equal(typeof exportsBlock.processApprovedFixes, 'function');
});

File: python/tests/approved_fixes_poll_contract_test.py
"""Contract test — dashboard-approved fix polling in patcherly_agent.py."""

from __future__ import annotations

from pathlib import Path


def test_patcherly_agent_polls_approved_and_applying() -> None:
    source = Path(__file__).resolve().parents[1].joinpath("patcherly_agent.py").read_text(encoding="utf-8")
    assert "_process_approved_fixes" in source
    assert "_apply_approved_fix_from_server" in source
    assert "_approved_apply_in_flight" in source
    assert '("approved", "applying")' in source or "('approved', 'applying')" in source
    assert "await self._process_approved_fixes()" in source
    # advanced_agent_testing keeps status=applying until test/results after approved apply
    apply_fn = source.split("async def _apply_approved_fix_from_server", 1)[1].split(
        "async def _process_approved_fixes", 1
    )[0]
    assert "await self._run_tests_and_report(error_id, apply_ok)" in apply_fn
    # post-apply-config/connector is unsigned JSON — must not require response HMAC
    pa_fn = source.split("async def _get_post_apply_connector_json", 1)[1].split(
        "async def ", 1
    )[0]
    assert "_verify_response_hmac" not in pa_fn
